packages/core/src/linker/query_list.ts

```diff
@@ -135,20 +151,27 @@ export class QueryList<T> implements Iterable<T> {
    * occurs.
    *
    * @param resultsTree The query results to store
+   * @param identityAccessor Optional functions for extracting stable object identity from a value
+   *     in the array.
    */
-  reset(resultsTree: Array<T|any[]>): void {
-    this._results = flatten(resultsTree);
-    (this as {dirty: boolean}).dirty = false;
-    (this as {length: number}).length = this._results.length;
-    (this as {last: T}).last = this._results[this.length - 1];
-    (this as {first: T}).first = this._results[0];
+  reset(resultsTree: Array<T|any[]>, identityAccessor?: (value: T) => unknown): void {
+    const self = this as QueryListInternal<T>;
+    (self as {dirty: boolean}).dirty = false;
+    const newResultFlat = flatten(resultsTree);
+    if (this._changesDetected = !arrayEquals(self._results, newResultFlat, identityAccessor)) {
+      self._results = newResultFlat;
+      self.length = newResultFlat.length;
+      self.last = newResultFlat[this.length - 1];
+      self.first = newResultFlat[0];
+    }
   }
 
   /**
    * Triggers a change event by emitting on the `changes` {@link EventEmitter}.
    */
   notifyOnChanges(): void {
-    (this.changes as EventEmitter<any>).emit(this);
+    if (this._changes && (this._emitDistinctChangesOnly ? this._changesDetected : true))
```

Comment:
This should be logically equivalent:

```suggestion
    if (this._changes && (this._changesDetected || !this._emitDistinctChangesOnly))
```

Reply:
fixed